pgrx/src/spi/client.rs

```diff
@@ -119,44 +128,20 @@ impl<'conn> SpiClient<'conn> {
     /// Returned name can be used with this method to retrieve the open cursor.
     ///
     /// See [`SpiCursor`] docs for usage details.
-    pub fn find_cursor(&self, name: &str) -> SpiResult<SpiCursor<'conn>> {
+    pub fn find_cursor(&self, name: &str) -> SpiResult<SpiCursor> {
         use pgrx_pg_sys::AsPgCStr;
 
         let ptr = NonNull::new(unsafe { pg_sys::SPI_cursor_find(name.as_pg_cstr()) })
             .ok_or(SpiError::CursorNotFound(name.to_string()))?;
-        Ok(SpiCursor { ptr, __marker: PhantomData })
+        Ok(SpiCursor { ptr, client: self })
     }
 }
 
-/// a struct to manage our SPI connection lifetime
-pub(super) struct SpiConnection(PhantomData<*mut ()>);
-
-impl SpiConnection {
-    /// Connect to Postgres' SPI system
-    pub(super) fn connect() -> SpiResult<Self> {
-        // connect to SPI
-        //
-        // SPI_connect() is documented as being able to return SPI_ERROR_CONNECT, so we have to
-        // assume it could.  The truth seems to be that it never actually does.  The one user
-        // of SpiConnection::connect() returns `spi::Result` anyways, so it's no big deal
-        Spi::check_status(unsafe { pg_sys::SPI_connect() })?;
-        Ok(SpiConnection(PhantomData))
-    }
-}
-
-impl Drop for SpiConnection {
-    /// when SpiConnection is dropped, we make sure to disconnect from SPI
+impl Drop for SpiClient {
     fn drop(&mut self) {
         // best efforts to disconnect from SPI
         // SPI_finish() would only complain if we hadn't previously called SPI_connect() and
-        // SpiConnection should prevent that from happening (assuming users don't go unsafe{})
+        // SpiClient will prevent that from happening (assuming users don't go unsafe{})
         Spi::check_status(unsafe { pg_sys::SPI_finish() }).ok();
     }
 }
```

Comment:
So the problem here is that now that it's on SpiClient, people can just run `mem::forget(client)` safely.

Reply:
The user only sees a `&mut client`.  Wouldn't `mem::forget()` on a reference be a noop?

Reply:
even better:

```
error[E0382]: borrow of moved value: `client`
   --> pgrx-tests/src/tests/spi_tests.rs:228:13
    |
226 |         Spi::connect(|client| {
    |                       ------ move occurs because `client` has type `&mut SpiClient`, which does not implement the `Copy` trait
227 |             std::mem::forget(client);
    |                              ------ value moved here
228 |             client.update("CREATE TABLE tests.cursor_table (id int)", None, None)?;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after move

```

Reply:
I didn't notice the change to `&mut SpiClient` from `SpiClient<'_>` when I said that. This seems like it could be an improvement.